### ISSUE TITLE
Breath cycle detector

### ISSUE THREAD

**Issue by cocov** [hev-sw/hev-analysis#9]
# Description of the issue

Several analysis rely on having identified the breath cycle. Start is the begining of inhale, stop the end of exhale. Flow rising up can be a good definition of inhale start. 

# Task
1. The function take the data from testchest or hev and from the flow return a set of time giving the starting time of breath. 
2. Could also return the repsiration per minute.

**Comment by jhaimber** [comment]
![ventilator](https://user-images.githubusercontent.com/63743093/79789714-ec177400-834a-11ea-9c9c-31eb08aabb6f.png)

Is this a sufficient precision for the breath cycle (the green plateau marks the detected breath)?

**Comment by cocov** [comment]
Hi Jakob, that's great!
The end of the cycle is actually the start of the next cycle rather than the moment the flow is back to 0. But the moment the flow is back to 0 is also usefull to have (the next 100ms or so can be used to extract the measured PEEP).
Can you give a bit of explanation on how you get the start? Do you do some baseline/noise evaluation, or is it a single threshold?

Could you try this on  https://lblogbook.cern.ch/HEV/40?hide=0,1,2,3,4, the one with "weak" in the name ? There the flow will start before the big jump. So that will be a good check.

@tevans1260 I did not though through the FSM stuff properly, when inhale will start will be the moment where we open the valve, but in the case of a spontaneous breath (like what you were looking at) the "start time" is not in the FSM, it is just before when flow start deviating from 0. Jakob code could be use for that if he is sensitive enough

**Comment by jhaimber** [comment]
If we define a cycle from teh start of a breath to the next, then there should be no problem.
![ventilator3](https://user-images.githubusercontent.com/63743093/79837955-9080d100-83b2-11ea-99bd-71fd6fa5ddf3.png)

**Comment by jhaimber** [comment]
To find the points, I impose four conditions on them.
1. The previous 5 points are within one std of the mean of all points
2.The derivation of the next 4 points is positiv
3.The mean derivation of the next 4 points is bigger than the mean-derivation of all points + std
4.The second point  after the selected point has to be bigger than the mean of all points + std

**Comment by cocov** [comment]
OK great! Can you try it on  https://lblogbook.cern.ch/HEV/40?hide=0,1,2,3,4, the one with "weak", zooming on one or 2 cycle such that we can see the start (it should correspond , on the pressure picture, to when there is an underpressure starting)?

**Comment by jhaimber** [comment]
![ventilator2](https://user-images.githubusercontent.com/63743093/79841673-0f2c3d00-83b8-11ea-82e1-06e40472c7ac.png)
The data from  https://lblogbook.cern.ch/HEV/40?hide=0,1,2,3,4,

**Comment by tevans1260** [comment]
okay - so for the spontaneous breath we will have to do something different, but for the non-spontaneous breath, could we check what you are getting @jhaimber with the analysis with the state codes? 

maybe i misunderstood what was said on Friday - but isn't the start of breath in the weakbreath here really at like ~ 25 s on the plot, i.e. when you go under pressure but the valve wasn't opening for this data so the rise gives you wrong point?

**Comment by cocov** [comment]
yes it is , the threshold on the flow should be much lower. Use the one from the Dp is it much less noisy than the flow.
Zoom also to see the little kick. 
I'll try to check if it is well there on those data (was taken at the begining).
